fix: `go-corset` running `test` in CI

DavePearce opened the pull request:
This disables the use of parallel forks in gradle unless this is
explicitly requested.  The issue is that parallel forks interferes with
go-corset which is highly concurrent.  Hence, running go-corset tasks in
parallel can (in some cases) lead to out-of-resource issues and workflow
tasks being cancelled unexpectedly.

However, since using parallel forks does improve the runtime for Rust
corset, this option is retained when specifically specified via the
UNIT_TESTS_PARALLELISM environment variable.

DavePearce committed:
Disable implicit parallel forks

This disables the use of parallel forks in gradle unless this is
explicitly requested.  The issue is that parallel forks interferes with
go-corset which is highly concurrent.  Hence, running go-corset tasks in
parallel can (in some cases) lead to out-of-resource issues and workflow
tasks being cancelled unexpectedly.

However, since using parallel forks does improve the runtime for Rust
corset, this option is retained when specifically specified via the
REPLAY/WEEKLY/REFERENCE_TESTS_PARALLELISM environment variables.

gauravahuja commented on gradle/tests.gradle:
```diff
@@ -118,7 +118,15 @@ tasks.withType(Test).configureEach {
 }
 
 tasks.test.configure {
-  maxParallelForks = Runtime.getRuntime().availableProcessors()
```
- The gradle task `test`, `./gradlew test`is only used for unit tests. Thus `maxParallelForks` is only used in unit tests.
- Replay, Weekly and Reference tests have their own gradle tasks that set the parallelism. You can set the parallelism for go corset workflows by either removing the `*_TESTS_PARALLELISM` env varaible or setting it to 1.
- You can introduce another variable for UNIT_TESTS_PARALLELISM that is disabled for go corset workflows

DavePearce replied:
So (surprisingly) the unit tests are the ones causing the problem (they are getting very expensive now).  We cannot have `maxParallelForks` being defined as above when `go-corset` is being used.  That's basically what I'm doing here.  We could have `UNIT_TESTS_PARALLELISM` and, when that is defined, set `maxParallelForks`.  That would work fine I guess, and is pretty comparable to what we've got here.

DavePearce replied:
But I do understand you are saying this is redundant:

```
   if ( ...
      System.getenv().containsKey("WEEKLY_TESTS_PARALLELISM") ||
      System.getenv().containsKey("REFERENCE_TESTS_PARALLELISM")) {
```

Because these keys are never used for the `test` task.

DavePearce replied:
Ok, I added `UNIT_TESTS_PARALLELISM`

DavePearce committed:
Support `UNIT_TESTS_PARALLELISM`

This allows the level of parallelism used for the unit tests to be
configured via an environment variable `UNIT_TESTS_PARALLELISM`.
Specifically, if this is left unset then no parallelism is used (which
is desirable when running go-corset).

gauravahuja commented on gradle/tests.gradle:
```diff
@@ -118,7 +118,13 @@ tasks.withType(Test).configureEach {
 
 tasks.test.configure {
-  maxParallelForks = Runtime.getRuntime().availableProcessors()
+  if (System.getenv().containsKey("UNIT_TESTS_PARALLELISM")) {
+    // Only set this if parallelism is explicitly requested.
+    // Specifically, because this option is not compatible with
+    // go-corset.
+    maxParallelForks = System.getenv().get("UNIT_TESTS_PARALLELISM").toInteger()
+    println "Enabled parallel forks."
+  }
```
```suggestion
       System.getenv().getOrDefault("UNIT_TESTS_PARALLELISM", "1").toInteger())
```
